xonsh/lexer.py

```diff
@@ -1,27 +1,333 @@
+    (tokenize.OP, '('): handle_lparen,
+    (tokenize.OP, ')'): handle_rparen,
+    (tokenize.OP, '{'): handle_lbrace,
+    (tokenize.OP, '}'): handle_rbrace,
+    (tokenize.OP, '['): handle_lbracket,
+    (tokenize.OP, ']'): handle_rbracket,
+    (tokenize.ERRORTOKEN, '$'): handle_dollar,
+    (tokenize.ERRORTOKEN, '`'): handle_backtick,
+    (tokenize.ERRORTOKEN, '?'): handle_question,
+    (tokenize.ERRORTOKEN, ' '): handle_error_space,
+}
+"""
+Mapping from ``tokenize`` tokens (or token types) to the proper function for
+generating PLY tokens from them.  In addition to yielding PLY tokens, these
+functions may manipulate the Lexer's state.
+"""
+
+
+def handle_token(state, token, stream):
+    """
+    General-purpose token handler.  Makes use of ``token_map`` or
+    ``special_map`` to yield one or more PLY tokens from the given input.
+
+    Parameters
+    ----------
+
+    state :
+        The current state of the lexer, including information about whether
+        we are in Python mode or subprocess mode, which changes the lexer's
+        behavior
+    token :
+        The token (from ``tokenize``) currently under consideration
+    stream :
+        A generator from which more tokens can be grabbed if necessary
+    """
+    typ = token.type
+    st = token.string
+    if not state['pymode'][-1]:
+        if state['last'] is not None and state['last'].end != token.start:
+            cur = token.start
+            old = state['last'].end
+            yield _new_token('WS', ' '*(cur[1]-old[1]), old)
+    if (typ, st) in token_map:
+        state['last'] = token
+        yield _new_token(token_map[(typ, st)], st, token.start)
+    elif typ in token_map:
+        state['last'] = token
+        yield _new_token(token_map[typ], st, token.start)
+    elif (typ, st) in special_handlers:
+        yield from special_handlers[(typ, st)](state, token, stream)
+    elif typ in special_handlers:
+        yield from special_handlers[typ](state, token, stream)
+    else:
+        m = "Unexpected token: {0}".format(token)
+        yield _new_token("ERRORTOKEN", m, token.start)
+
+
+def get_tokens(s):
+    """
+    Given a string containing xonsh code, generates a stream of relevant PLY
+    tokens using ``handle_token``.
+    """
+    tokstream = tokenize.tokenize(BytesIO(s.encode('utf-8')).readline)
+    state = {'indents': [0], 'pymode': [True], 'last': None}
+    while True:
+        try:
+            token = next(tokstream)
+            yield from handle_token(state, token, tokstream)
+        except StopIteration:
+            return
+        except tokenize.TokenError as e:
+            # this is recoverable in single-line mode (from the shell)
+            # (e.g., EOF while scanning string literal)
+            yield _new_token('ERRORTOKEN', e.args[0], (0, 0))
+            return
+        except IndentationError as e:
+            # this is never recoverable
+            yield _new_token('ERRORTOKEN', e, (0, 0))
+            return
```

Comment:
I didn't think that you could have both returns and yields.  In any event, it is probably better to use `break` here.